os/fs/smartfs/smartfs_smart.c

```diff
@@ -1696,13 +1716,25 @@ static int smartfs_mkdir(struct inode *mountpt, const char *relpath, mode_t mode
 
 		/* Create the directory */
 
-		ret = smartfs_createentry(fs, parentdirsector, filename,
-								  SMARTFS_DIRENT_TYPE_DIR, mode, &entry, 0xFFFF, NULL);
-		if (ret != OK) {
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+		ret = smartfs_find_firstinvalidentry(fs, parentdirsector, &invalidentry);
+		if (ret == OK) {
+			invalidentry.firstsector = SMARTFS_ERASEDSTATE_16BIT;
+			ret = smartfs_replaceentry(fs, invalidentry, filename, SMARTFS_DIRENT_TYPE_DIR, mode, &entry, NULL);
+			if (ret != OK) {
+				goto errout_with_semaphore;
+			}
+		} else if (ret < 0 && ret != -ENOENT) {
```

Comment:
-ENOENT should not come to as a result.
1. Find first invalid entry.
2. If there's only valid entry, return first empty entry. (we can call createentry there?)
So I don't think it returns -ENOENT ...

Reply:
Function find_firstinvalidentry() will return -ENOENT when there is no invalidated entry in any of the chained sectors.

However, now the logic has been improved. This function also checks for free entries along the way so if no invalidated entry is found, it can return the first free entry found.
Still if no free or invalidated entry is found it will return -ENOENT.